### ISSUE TITLE
New file boilerplate

### ISSUE THREAD

**Issue by jon-bondy** [openscad/openscad#4458]
I start all of my OpenSCAD programs the same way, establishing some constants (like "inches = 25.4;") and importing/using some files.  I type this in by hand, over and over and over again.

It would be nice if, when OS created a New file, and if a file like "Default.scad" is present in (say) the Library folder, then the New file is filled with the contents of the Default.scad file.  Since this file will not exist on older installations, this new feature will only affect those who wish to use it going forward.

Thanks for considering this.

Jon

**Comment by jordanbrown0** [comment]
Might be worthwhile, but note that you can reduce your per-file manual work to
```
include <Default.scad>
```
Note that then if you want to make a change to your boilerplate, it immediately affects all of your models.  That might be good, or might be bad.

**Comment by t-paul** [comment]
You can store your own code snippets in the config folder, e.g. on Linux `.config/OpenSCAD/templates` and insert those via `ALT`+`INS` or Context-Menu->Insert Template.

Example files: https://github.com/openscad/openscad/tree/master/templates

**Comment by jon-bondy** [comment]
Nice.  I had no idea this was possible.  But I like my idea better.  Of 
course I do!

Thanks!

**Comment by t-paul** [comment]
Yes, I think we can keep this open as separate feature. However I would not go with a single fixed file, but instead something like "New from template" which can select from files either shipped or user provided.

**Comment by jon-bondy** [comment]
If I edit a Template and save it and then go into OpenSCAD and insert it, the edits are not present.  Are the Templates loaded at program startup, or when the popup menu is created?  It would seem that the latter would be preferred.

**Comment by jordanbrown0** [comment]
It appears that they are loaded when the editor tab is created.   Agreed that it would be better if changes after that took immediate effect.

But let's keep this enhancement request on topic:  new-file boilerplate.  Hiding these two comments; please continue any comments about templates per se in another issue or on the mailing list.

**Comment by bitbasher** [comment]
i suggest this feature have these parts : 
1. improve the documentation of the template feature as it is now - it needs its own section
2. add the OpenSCAD folder at installation to the $HOME/.config folder with subfolders Backups, Libraries, and Templates .. which the docs say should already be happening but is not .. at least not in Windows nor in Mac OS
3. add the option of a template to be automatically loaded into a new Editor Tab in the preferences
4. make the default name of this template something like "newFileDefaultTemplate
5." and that be a preferences setting
6. add a menu item Edit > MakeTemplate that will take the current selection in the Editor Tab, or the entire text of the script, into a new edit tab as a .json file that can then be edited as a template. Then be able to save that template.json file into the users $HOME/.config location 

there might be some additional features discovered .. but being able to use templates in the editor is huge

**Comment by jordanbrown0** [comment]
Let's keep it to one feature/bug per issue.  This one would be your number 3, I think.

**Comment by bitbasher** [comment]
> Let's keep it to one feature/bug per issue. This one would be your number 3, I think.

makes sense 
i have made a start on #1 with #6281 and a revision to the [User Interface page](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/The_OpenSCAD_User_Interface#Editor_Templates) to document templates

for this issue i will focus on #3 .. That when File > New is done that IF there is a file
  [User_Config_Path]/templates/LoadOnNew-template.json
that file will be pasted into the editor as text and will not be parsed as code

hummm .. and i think i should leave it off of the templates list popup from the context menu .. opinions?  The user would not need to re-paste the auto-loaded template would he?

**Comment by jordanbrown0** [comment]
I agree with what Torsten said above:
>However I would not go with a single fixed file, but instead something like "New from template" which can select from files either shipped or user provided.

Note:  new-from-template might not be exactly the same as inserting a template.  It might be more "Open existing file as new file" - that is, it reads the existing file, but the top of the window says "Untitled.scad".

**Comment by jon-bondy** [comment]
I, on the other hand, would like things to be really simple for me.  I do NOT want to have to select a template.  I want a template to be used automatically, if it is present.

We can have both approaches without any conflict.  New would use the automatic approach and New With Template would use Jordan's approach.

Jon

On 10/14/2025 10:50 PM, Jordan Brown wrote:
[https://avatars.githubusercontent.com/u/31949071?s=20&v=4]jordanbrown0 left a comment (openscad/openscad#4458)<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_openscad_openscad_issues_4458-23issuecomment-2D3404316330&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=AsrE-c7ZR7B2Kyr3qgfvvppkCEBVsNmwEMndcrRSuOI&m=nuwRXV7NGxe3UCddmm10svjaSPnx7rvfuUH8o2O4mc-xf_S355b2o6jVCQ-h6u0y&s=DMBD2ViMZiaAJyDdd-o-lBj952oZdc8k41soU0ucW4A&e=>

I agree with what Torsten said above:

However I would not go with a single fixed file, but instead something like "New from template" which can select from files either shipped or user provided.

Note: new-from-template might not be exactly the same as inserting a template. It might be more "Open existing file as new file" - that is, it reads the existing file, but the top of the window says "Untitled.scad".

**Comment by bitbasher** [comment]
i think we can do both 

Part 1
the user can create a specially named file like /User/vulcan_/.config/OpenSCAD/templates/UserDefinedStartTemplate.json
It will be easy to notice this special file as the templates are being loaded by Scintilla as it starts up .. it could even have a preference checkbox to enable the auto-loading template ( .. and a text field to specify the filename? possible but is it too much? )

should that be Enabled by Default ? or require the user to enable it explicitly?

Part 2
a new menu File > New From Template .. pop-up, select from list, and paste it into place once the editor tab is open

this is not going to be much use if templates are simple strings that need explicit \n \t for formatting .. i think i would want to add the possibility of multi-line text templates.  If that is the case .. would i want to allow include <> and use <> to operate when pasted into a new editor tab .. or should they be excluded from .scad text blocks?

this would also mean having a section on the start up splash window for templates .. either widen the splash to have space for three selection lists .. or just have one area for selections with three tabs?

**Comment by jon-bondy** [comment]
I suppose we could add a new field to the JSON record called "tag" which could be blank (meaning put this in the list when the user attempts an Insert Template), or "new" (meaning put this in the list when New From Template is used) or "default" (meaning insert this every time a new file is created.  Then the user is in complete control, assuming they are able to locate and edit the files.

Jon


On 10/15/2025 5:23 PM, Jeff Hayes wrote:
[https://avatars.githubusercontent.com/u/430590?s=20&v=4]bitbasher left a comment (openscad/openscad#4458)<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_openscad_openscad_issues_4458-23issuecomment-2D3408350640&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=AsrE-c7ZR7B2Kyr3qgfvvppkCEBVsNmwEMndcrRSuOI&m=oI-q6xkUF0NDQJsgyhqUr0VfmPV20xzw_eSoiE0BgvQacM_6LYXCDNhkKTicqnI7&s=MrjgJXpo58fnHlEpol-8_u5QiKSlsjP--zNfkPCcJtU&e=>

i think we can do both

Part 1
the user can create a specially named file like /User/vulcan_/.config/OpenSCAD/templates/UserDefinedStartTemplate.json
It will be easy to notice this special file as the templates are being loaded by Scintilla as it starts up .. it could even have a preference checkbox to enable the auto-loading template ( .. and a text field to specify the filename? possible but is it too much? )

should that be Enabled by Default ? or require the user to enable it explicitly?

Part 2
a new menu File > New From Template .. pop-up, select from list, and paste it into place once the editor tab is open

this is not going to be much use if templates are simple strings that need explicit \n \t for formatting .. i think i would want to add the possibility of multi-line text templates. If that is the case .. would i want to allow include <> and use <> to operate when pasted into a new editor tab .. or should they be excluded from .scad text blocks?

this would also mean having a section on the start up splash window for templates .. either widen the splash to have space for three selection lists .. or just have one area for selections with three tabs?

**Comment by bitbasher** [comment]
> I suppose we could add a new field to the JSON record called "tag" which could be blank (meaning put this in the list when the user attempts an Insert Template), or "new" (meaning put this in the list when New From Template is used) or "default" (meaning insert this every time a new file is created.  Then the user is in complete control, assuming they are able to locate and edit the files.

The way they are loaded would make that possible .. the editor reads all of the json's in a folder and adds them to a boost property list .. i would separate the template "types" into separate lists just to make it easier to fill the lists in the UI without having to process them looking for the right tag .. but yeah, that is the thing to do

thx Jon

**Comment by bitbasher** [comment]
ok, i am able to get back to this now
